pennylane/measurements/expval.py

```diff
@@ -52,4 +56,65 @@ def circuit(x):
     if not op.is_hermitian:
         warnings.warn(f"{op.name} might not be hermitian.")
 
-    return MeasurementProcess(Expectation, obs=op)
+    return _Expectation(Expectation, obs=op)
+
+
+# TODO: Make public when removing the ObservableReturnTypes enum
+class _Expectation(SampleMeasurement, StateMeasurement):
+    """Measurement process that computes the probability of each computational basis state."""
+
+    def process_samples(
+        self,
+        samples: Sequence[complex],
+        wire_order: Wires,
+        shot_range: Tuple[int] = None,
+        bin_size: int = None,
+    ):
+        if isinstance(self.obs, Projector):
+            # branch specifically to handle the projector observable
+            idx = int("".join(str(i) for i in self.obs.parameters[0]), 2)
+            probs = qml.probs(wires=self.wires).process_samples(
+                samples=samples, wire_order=wire_order, shot_range=shot_range, bin_size=bin_size
+            )
+            return probs[idx]
+        # estimate the ev
+        samples = qml.sample(op=self.obs).process_samples(
+            samples=samples, wire_order=wire_order, shot_range=shot_range, bin_size=bin_size
+        )
+        # With broadcasting, we want to take the mean over axis 1, which is the -1st/-2nd with/
+        # without bin_size. Without broadcasting, axis 0 is the -1st/-2nd with/without bin_size
+        axis = -1 if bin_size is None else -2
+        # TODO: do we need to squeeze here? Maybe remove with new return types
+        return qml.math.squeeze(qml.math.mean(samples, axis=axis))
+
+    def process_state(self, state: Sequence[complex], wire_order: Wires):
+        if isinstance(self.obs, Projector):
+            # branch specifically to handle the projector observable
+            idx = int("".join(str(i) for i in self.obs.parameters[0]), 2)
+            probs = qml.probs(wires=self.wires).process_state(state=state, wire_order=wire_order)
+            return probs[idx]
+        eigvals = qml.math.asarray(self.obs.eigvals(), dtype="float64")
```

Comment:
Do we need to handle the case where the eigenvalues aren't defined?

Reply:
We can add the try/except logic and raise a more specific error (like in `QubitDevice`):
```python
try:
    eigvals = self._asarray(observable.eigvals(), dtype=self.R_DTYPE)
except qml.operation.EigvalsUndefinedError as e:
    raise qml.operation.EigvalsUndefinedError(
        f"Cannot compute analytic expectations of {observable.name}."
    ) from e
```

However when trying to test this I realised there is not a single gate in PL without eigvals defined. Then I thought that users might use custom operators, but I also realised that each device supports specific observables, thus an error would be raised if an unrecognised observable is used.

@albi3ro do you think we should check for `EigvalsUndefinedError` here?